stuffer/s2n_stuffer.c

```diff
@@ -52,7 +52,14 @@ bool s2n_stuffer_reservation_is_valid(const struct s2n_stuffer_reservation* rese
         return false;
     }
     const struct s2n_stuffer stuffer_obj = *(reserve_obj.stuffer);
-    return S2N_MEM_IS_WRITABLE(stuffer_obj.blob.data + reserve_obj.write_cursor, reserve_obj.length);
+    if (stuffer_obj.blob.size < reserve_obj.length) {
+        return false;
+    }
+    return S2N_IMPLIES(
+        reserve_obj.length > 0,
+        (reserve_obj.write_cursor < stuffer_obj.blob.size &&
```

Comment:
This seems looser than it needs to be: if reserve_obj.length > 0, then we should be able to assert reserve_obj.write_cursor < suffer_obj.write_cursor, which thanks to s2n_stuffer_is_valid is already shown to be <= stuffer_obj.blob.size.

Reply:
Hi,

Thanks for the review. I had committed this change to fix a failing proof.
I see that `s2n_stuffer_is_valid` (called at line 51) does ensure `suffer_obj.write_cursor <= stuffer_obj.blob.size`, but I don't see how we would be able to assert `reserve_obj.write_cursor < suffer_obj.write_cursor` ...

Reply:
If reserve_obj.length > 0, then two things should be true for the reservation object to be valid: its write_cursor should indicate the start of the reserved memory, and its stuffer_obj should have had its own write_cursor skipped ahead by at least the length of the reservation, or further if there was a subsequent write.

Reply:
To keep things as nondeterministic as possible, I think we would want the `reserve_obj.write_cursor` to point so "some" valid memory, not only the start of it.
Similarly, the `stuffer_object.write_cursor` could point to any valid memory, but we should make sure that the `stuffer_object.blob` is at least as big as the `reserve_obj.length`.

Reply:
Thanks @agray256. I have updated this check as per our discussion.